CHANGELOG.md

```diff
@@ -24,6 +24,11 @@ Usage:
 
 ### 1.9.1 (2022-09-14)
 * Fix #1747: Apply service doesn't attempt to create OpenShift Projects in Kubernetes clusters
+* Fix #1758: ServiceAccount binding not being generated in Deployment
+
+_**Note**_:
+- `resources` > `serviceAccount` configuration would no longer create a new ServiceAccount. It would only add ServiceAccount reference to Deployment, Job etc. Users are advised to use `resources` > `serviceAccounts` configuration option in order to generate and add ServiceAccounts.
```

Comment:
these two configuration options are not equivalent. `serviceAccount` adds `.spec.template.spec.serviceAccountName` in default controller. `serviceAccounts` adds `.spec.template.spec.serviceAccountName` only to controller name as specified in `deploymentRef` configuration, it also creates the ServiceAccount if it doesn't already exist.

Reply:
And you don't see how that might be extremely confusing?

Reply:
https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin#serviceaccount-generation

Reply:
We should probably discuss this properly.

If we define `<serviceAccounts>`, will this also bind one of the (first) created service accounts to the controller besides creating the resources?

Reply:
Cool, this makes things much easier.

Considering we modify the `ServiceAccount` class to include a new field `generate` which defaults to `true`. And another `boolean` field `bindToAllControllers` that defaults to `false`

The following configurations would be equivalent
```xml
<configuration>
    <resources>
      <serviceAccounts>
        <serviceAccount>
          <name>my-existing-serviceaccount</name>
          <bindToAllControllers>true</bindToAllControllers>
          <generate>false</generate>
        </serviceAccount>
      </serviceAccounts>
    </resources>
</configuration>
```
```xml
<configuration>
    <resources>
      <serviceAccount>my-existing-serviceaccount</serviceAccount>
    </resources>
</configuration>
```

right?

If so, this would simplify the configuration and lead to a single point of entry for the enrichers.

The only step would be to configure ResourceConfig deserialization to set a single service account string as an item in the serviceAccounts list with the overridden defaults.

Reply:
`serviceAccount` configuration option doesn't bind to all controllers. Only the ones that are generated by jkube

Reply:
I think the reason this note is still there is because I moved ServiceAccount related code (setting `.spec.serviceAccountName` in Handlers to ServiceAccountEnricher. I think it was a side effect of controller and ServiceAccount enrichers' execution order.

Earlier when `<serviceAccount>` option was set, ControllerEnrichers set ServiceAccount in PodTemplateSpec:
https://github.com/eclipse/jkube/blob/b212dff54e5d6c464c5fc2a8bd4725414aa74f03/jkube-kit/enricher/api/src/main/java/org/eclipse/jkube/kit/enricher/handler/PodTemplateHandler.java#L56

ServiceAccountEnricher (which executes after ControllerEnrichers) sees `.spec.serviceAccountName` field already set and creates ServiceAccount :
https://github.com/eclipse/jkube/blob/b212dff54e5d6c464c5fc2a8bd4725414aa74f03/jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/ServiceAccountEnricher.java#L65

In this PR, I moved all ServiceAccount-related logic from handlers to ServiceAccountEnricher. Now it's handled here along with `serviveAccounts` config.

Reply:
This note is outdated now. I've removed this note from CHANGELOG.